### PR TITLE
New Visual C# Code Snippets Cheat Sheet

### PR DESCRIPTION
**Pull request by Khalimov**
https://duck.co/ia/view/c_sharp_snippets_cheat_sheet

### CODE REVIEW

**Comment by daxtheduck** [comment]
@Khalimov Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

While the team and the community review your submission, we'd like to make sure you look good! If you'd like a free DuckDuckGo tshirt, please send over the following info to merch@duckduckgo.com, and be sure to put "Dev Merch" in the subject line:

Name:
Street Address:
City:
State / Province:
Country:
Phone (Non-US Orders only): 
T-shirt size:
GitHub ID:
Link to this Pull Request:

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: https://duck.co/duckduckhack/submission_and_review

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by moollaza** [comment]
@Khalimov Thanks for contributing and welcome to DuckDuckHack!

Can you please change the id of the IA page, and cheat sheet to include `cheat_sheet` at the end?

I think an ID like `c_sharp_cheat_sheet` would be best.

File: share/goodie/cheat_sheets/json/visual-studio-c#-snippets.json [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        "sourceUrl": "https://msdn.microsoft.com/en-us/library/z41h7fat.aspx"
+    },
+    "aliases": [
+        "c# snippets", "visual studio c# snippets", "visual c# snippets"  
+    ],
+    "template_type": "code",
+    "section_order": ["Declarations", "Blocks", "Loops", "Directives", "Others"],
+    "sections": {
+        "Directives": [{
+            "val": "Creates a #if directive and a #endif directive",
```
Comment (GuiltyDolphin) [inline share/goodie/cheat_sheets/json/visual-studio-c#-snippets.json]: @Khalimov, I'm curious - how do you pronounce `#if` and `#endif` in your head? I pronounce them as just `if` and `endif`, so the `a` (rather than `an`) feels a bit weird - what is it for you?

Comment (moollaza) [inline share/goodie/cheat_sheets/json/visual-studio-c#-snippets.json]: agreed! "Creates an #if directive" looks/sounds correct to me

File: share/goodie/cheat_sheets/json/c-sharp-snippets.json [hunk 2]
```diff
@@ -0,0 +1,132 @@
+{
+    "id": "c_sharp_snippets_cheat_sheet",
+    "name": "Visual Studio C# Snippets Cheat Sheet",
```
Comment (jophab) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: @Khalimov  I think we can remove the word cheat sheet from name

Comment (Khalimov) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: @moollaza Should I edit name?

Comment (GuiltyDolphin) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: @Khalimov I would say `Visual Studio C# Snippets` would be fine for the name - as that is fully descriptive of what the 'cheat sheet' is presenting. Having `Cheat Sheet` on the end is rather redundant.

Comment (Khalimov) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: @GuiltyDolphin sorry, I'm new in this, name in json should match with IA page name or not?

Comment (GuiltyDolphin) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: @Khalimov It's fine :) Always happy to see new contributers!

The `"name'` attribute is what is displayed the top of the cheat sheet.

For example, in the Wunderlist cheat sheet, `Wunderlist` (big, bold, at the top) is the `name`.
screenshot from 2015-12-30 12 35 42

So the `name` doesn't *need* to match anything. 

I'd recommend having a look at [the docs](http://docs.duckduckhack.com/frontend-reference/cheat-sheet-reference.html) and scroll down a bit to the `JSON Cheat Sheet Reference` which should clear up a few things.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=c# snippets cheat sheet)
                
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/c_sharp_snippets_cheat_sheet_mobile.png)

File: share/goodie/cheat_sheets/json/c-sharp-snippets.json [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    "section_order": ["Declarations", "Blocks", "Loops", "Directives", "Others"],
+    "sections": {
+        "Directives": [{
+            "val": "Creates a #if directive and a #endif directive",
+            "key": "#if"
+        }, {
+            "val": "Creates a #region directive and a #endregion directive",
+            "key": "#region"
+        }, {
+            "val": "Creates a using directive",
```
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: It might  be better off with a description like `imports the namespace into the current file` or similiar

File: share/goodie/cheat_sheets/json/c-sharp-snippets.json [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            "key": "using"
+        }],
+        "Declarations": [{
+            "val": "Creates a declaration for a class that derives from Attribute",
+            "key": "attribute"
```
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: as far as I know, attribute isn't a keyword. It's used when declaring a class to be an attribute as part of the name. Or is there another use I'm not aware of?

File: share/goodie/cheat_sheets/json/c-sharp-snippets.json [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            "key": "enum"
+        }, {
+            "val": "Creates a method declaration that overrides the Equals method defined in the Object class",
+            "key": "equals"
+        }, {
+            "val": "Creates a declaration for a class that derives from an exception (Exception by default)",
+            "key": "exception"
+        }, {
+            "val": "Creates an indexer declaration",
+            "key": "indexer"
```
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: Similar here, indexers are basically operator overloads for `[]`; I don't recall the actual string `indexer` being used?

File: share/goodie/cheat_sheets/json/c-sharp-snippets.json [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            "key": "sim"
+        }, {
+            "val": "Creates a struct declaration",
+            "key": "struct"
+        }, {
+            "val": "Creates a static void Main method declaration",
+            "key": "svm"
+        }],
+        "Blocks": [{
+            "val": "Creates a checked block",
```
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: It's probably worth defining what `checked` means: `creates a block in which integer arithmetic is checked for overflows`; a bit wordy but something like that?

**Comment by mintsoft** [comment]
@Khalimov Apologies, I've just realised that this is targeted at snippets (duh!) rather than keywords etc; so you can dis-regard most of my comments!

**Comment by moollaza** [comment]
@Khalimov apologies for the delay! This LGTM :+1:

Thanks a lot for contributing (we really appreciate it :smiley:). We invite you to stick around and comment/help on other PR's & Issues or even submit an Instant Answer!

If you need ideas about how to contribute next, please feel free to reach out to me on Slack! I'm happy to chat about how you can help :+1:

File: share/goodie/cheat_sheets/json/c-sharp-snippets.json [hunk 7]
```diff
@@ -0,0 +1,132 @@
+{
+    "id": "c_sharp_snippets_cheat_sheet",
+    "name": "Visual Studio C# Snippets",
+    "description": "Code Snippets",
```
Comment (moollaza) [inline share/goodie/cheat_sheets/json/c-sharp-snippets.json]: nit: we should make this a little more descriptive, or remove it

**Comment by moollaza** [comment]
@Khalimov thanks again for creating the C# Cheat Sheet! You did a great job. We at DuckDuckGo, the DuckDuckHack community, and all the people who will encounter this Instant Answer really appreciate it.

We could really use your help on something else now. We want to increase our Instant Answer coverage for as many programming languages and platforms as possible. We want a C# Package Search Instant Answer and I think you'd similarly do a great job with building that. Please update the [Programming IA Coverage table](https://github.com/duckduckgo/duckduckgo/wiki/Programming-IA-Coverage) (mark it with a 'd') if you'd be willing to take it up!

I think it is a good fit for you, but I'm of course not entirely sure, so I'd love to chat if you're around this week or next. Please message me on slack (@moollaza) or send me an email (moollaza@duckduckgo.com) if you'd be interested in talking. 

Thanks again!
